### ISSUE TITLE
regularized_qr's backward is O(1) wrong on complex input at every shape (real branch only; rejected in #913)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#917]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

`regularized_qr`'s backward is the *real* branch only. On complex input it returns a finite, plausible gradient that is **wrong by O(1) relative error — for every shape**, not just the wide one. It is currently rejected outright (#913) rather than left silently wrong; this issue is to implement it properly.

Found by Codex review on #913, which caught the wide case. Measuring it showed square and tall were already affected.

## Measured

Against `jax.vjp(jnp.linalg.qr, .)`, which handles complex correctly and is therefore a valid oracle:

```
   shape    kind     dtype   rel err vs jax.vjp
     4x4  square      real            1.024e-16
     4x4  square   complex            1.248e+00  <-- WRONG
     5x3    tall      real            1.157e-16
     5x3    tall   complex            1.427e+00  <-- WRONG
    12x6    tall      real            3.926e-16
    12x6    tall   complex            1.151e+00  <-- WRONG
     3x5    WIDE      real            7.089e-17
     3x5    WIDE   complex            1.659e+00  <-- WRONG
     4x8    WIDE      real            9.635e-16
     4x8    WIDE   complex            1.906e+00  <-- WRONG
   16x32    WIDE      real            9.203e-16
   16x32    WIDE   complex            1.608e+00  <-- WRONG
```

Real is exact to machine precision at every shape. Complex is off by 115%–191%. A gradient that wrong is not an approximation — it does not point in the right direction.

## Why it is wrong

`_ad_primitives.py:578`, in `regularized_qr`'s own docstring:

> This is the *real* branch only: CTM projector matrices are real, so the complex diagonal correction in JAX's thin-QR JVP is intentionally omitted.

The omission is deliberate and documented. **The premise is what has gone stale.** The QR projector's gauge fix handles complex explicitly (`_ctm_projector.py`, in the `regularized_qr` call site):

> For complex matrices, `diag_R >= 0` is undefined; rotate by the phase of diag(R) so diag(R) becomes real and non-negative. This reduces to the ±1 sign convention in the real case.

So the projector is built to accept complex, while the rule differentiating it assumes real. Nothing connects the two claims, and nothing tested the combination — every pre-existing test in `tests/test_regularized_qr.py` uses real input.

JAX's thin-QR JVP carries the correction this omits:

```python
do = qt_dx_rinv_lower - _H(qt_dx_rinv_lower)
# The following correction is necessary for complex inputs
do = do + I * (qt_dx_rinv - jnp.real(qt_dx_rinv))
```

## What I tried, so the next person does not repeat it

The textbook Seeger-style form:

```python
X   = R @ H(dR) - H(dQ) @ Q
low = jnp.tril(X, -1)
cop = low + H(low) + jnp.diag(jnp.real(jnp.diag(X)))   # real part on the diagonal
Abar = dQ + Q @ cop
dM   = H(solve_triangular(R, H(Abar), lower=False))
```

**This does not match JAX either** — 1.169 / 1.455 / 1.594 relative error on complex 4x4 / 5x3 / 12x6, while staying exact on real. So the fix is not simply "take the real part of the diagonal".

The likely reason is convention: **JAX pairs complex cotangents unconjugated** (the VJP contracts `Re sum(g * dz)`, not `Re sum(conj(g) * dz)`), whereas the textbook QR-VJP derivations assume the conjugated Wirtinger pairing. A rule transcribed from a paper will therefore disagree with `jax.vjp` by exactly this kind of O(1) factor even when the mathematics is right in its own convention. Getting this wrong does not crash — it returns a plausible number — which is why this needs to be verified numerically rather than reviewed by eye.

## What a fix needs

1. Implement the complex branch, deriving in **JAX's** cotangent convention rather than transcribing a paper's.
2. Verify against `jax.vjp(jnp.linalg.qr, .)` to ~1e-14 relative for **complex** square, tall **and** wide — the existing parity tests cover real only, which is the hole that let this sit.
3. Keep the `diag(R)` floor working in the complex branch. That floor is `regularized_qr`'s entire reason to exist over `jnp.linalg.qr`, and parity tests use well-conditioned inputs where it never activates, so it needs its own finiteness test at complex rank-deficiency — with an assertion that raw `jnp.linalg.qr` really does go non-finite there, so the guard is not vacuous.
4. Remove the `NotImplementedError` added in #913 and replace it with the parity tests.

## Current behaviour (#913)

`_regularized_qr_bwd` raises `NotImplementedError` on complex input, naming the measured error magnitudes and pointing at `projector_method='svd'`/`'eigh'`.

That is a deliberate scope increase over what #913 set out to do. #913 fixed a *crash* on wide real matrices; without the guard it would also have converted the complex wide case from a loud shape error into a silent O(1)-wrong gradient. Rejecting square and tall complex as well is the same argument applied to a pre-existing hole: those were never right, and a caller has no way to tell. Failing loudly is strictly better than a 120%-wrong optimization direction.

The cost is that complex + `projector_method="qr"` now raises where it previously ran. It previously ran *wrongly*, so no correct result is lost — but if anyone has a complex QR AD run in flight, this is the change that will stop it, and the fix here is what unblocks them.

## Related

#913 (added the rejection; fixed the wide-real crash), #912 (the wide crash), #570 (the spike that established the real-only rule and verified it "for square AND tall real matrices"), #858 (whose QR adjoint question #913 answered, on real tensors).

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Fixed in #913 — by not deriving it

Closing approach: the rejection this issue documented has been **replaced by a working complex branch**, so `projector_method="qr"` now differentiates complex input correctly rather than raising.

The insight is that hand-deriving was the wrong move at every step. What `regularized_qr` actually adds over `jnp.linalg.qr` is **only** the `diag(R)` floor — everything else is supposed to be JAX's rule. So apply the floor and ask JAX:

```python
R_reg = floor(diag(R))
M_reg = Q @ R_reg
_, vjp_qr = jax.vjp(jnp.linalg.qr, M_reg)
return vjp_qr((dQ, dR))
```

Correctness splits cleanly along whether the floor engages:

* **Well-conditioned** — `R_reg == R`, so `M_reg == M` and this is literally `jax.vjp(jnp.linalg.qr, M)`. Parity holds **by construction rather than by derivation**, for real and complex alike. The convention question that defeated the hand-written attempt never arises, because JAX is answering in its own convention.
* **Rank-deficient** — the true derivative does not exist (that is the whole reason a floor exists), and this evaluates JAX's rule at a nearby non-singular factorization. That *is* the regularization, now stated directly instead of smuggled into a triangular solve.

## Measured

Parity vs `jax.vjp(jnp.linalg.qr, .)`, ~1e-15 relative across **square / tall / wide x real / complex**, including the shapes the C4v projector produces (`16x32`, `32x64`, `48x96`). Compare the numbers in the issue body: 1.248 / 1.427 / 1.659 / 1.608 before.

## This also retired the #912 fix

The hand-written wide branch is gone — JAX handles wide natively, so #912's shape bug cannot recur in a form that needs its own code path. Net **-52 lines**: wide branch, square/tall helper, and the complex rejection all collapse into delegation.

## Mutation-checked, because "delegate to JAX" can silently drop either guarantee

| mutation | result |
|---|---|
| force real cotangents (simulates the old real branch) | **6 failed** — all complex parity cases |
| remove the `diag(R)` floor | **3 failed** — all finiteness cases |

So both properties are load-bearing rather than incidental.

## Cost

One extra QR of `M_reg` in the backward, replacing a triangular solve. For CTM projector sizes that is not a concern, but it is a real change in the backward's cost and worth knowing.

## A pre-existing vacuous guard, found by the mutation and deliberately left alone

`test_regularized_qr_backward_finite_at_rank_deficiency` (real, square) does **not** fail when the floor is removed. Its comment says *"raw QR VJP would NaN"* — raw JAX is in fact finite on that input, so the floor never engages there and the test has been passing for free. That predates all of this and is not touched here; the new finiteness tests assert non-vacuity explicitly (they check raw JAX really does go non-finite). Worth a follow-up to give that test an input that actually exercises what it claims.
